### PR TITLE
Add task infrastructure component

### PR DESCRIPTION
**Pull request by tongwu-msft**
Add task infrastructure component

## Description
Add task infrastructure component. 
 - Task hosting: host task execution environment.
 - Task consumer: client to pull/keepalive/complete task.
 - Task manager: client to create/cancel/get task.

This component is to handle different async task execution. 
- Support to easy create new kind of task. 
- General task management component to manage task: pull/complete/keepalive operations.

__Sample TaskHosting Wrapper in backend service__
```
public class TaskHostingBackgroundService : BackgroundService
    {
        private readonly Func<IScoped<TaskHosting>> _taskHostingFactory;

        protected override async Task ExecuteAsync(CancellationToken stoppingToken)
        {
            using (IScoped<TaskHosting> taskHosting = _taskHostingFactory())
            {
                await taskHosting.Value.StartAsync(CancellationTokenSource.CreateLinkedTokenSource(stoppingToken));
            }
        }
    }
```

__Sample Task Factory__
```
public class TaskFactory : ITaskFactory
    {
        public ITask Create(TaskInfo taskInfo)
        {
            if (taskInfo.TaskTypeId == BulkImportDataProcessingTask.BulkImportDataProcessingTaskTypeId)
            {
                IContextUpdater contextUpdater = _contextUpdaterFactory.CreateContextUpdater(taskInfo.TaskId, taskInfo.RunId);
                BulkImportDataProcessingInputData inputData = JsonConvert.DeserializeObject<BulkImportDataProcessingInputData>(taskInfo.InputData);
                BulkImportProgress bulkImportProgress = string.IsNullOrEmpty(taskInfo.Context) ? new BulkImportProgress() : JsonConvert.DeserializeObject<BulkImportProgress>(taskInfo.Context);
                return new BulkImportDataProcessingTask(
                    inputData,
                    bulkImportProgress,
                    _fhirDataBulkOperation,
                    contextUpdater,
                    _resourceLoader,
                    _importErrorUploader,
                    _rawResourceProcessor,
                    _bulkImporter,
                    _contextAccessor,
                    _loggerFactory);
            }

            return null;
        }
    }
```

__Task interface__
```
    public interface ITask : IDisposable
    {
        public string RunId { get; set; }

        public Task<TaskResultData> ExecuteAsync();

        public void Cancel();

        public bool IsCancelling();
    }
```

## Related issues
[#81238](https://microsofthealth.visualstudio.com/Health/_workitems/edit/81238)

## Testing
Add unit tests for all components. Integration and e2e test would be added when it integrated with fhir service.

## FHIR Team Checklist
- [x] **Update the title** of the PR to be succinct and less than 50 characters
- [x] **Add a milestone** to the PR for the sprint that it is merged (i.e. add S47)
- [x] Tag the PR with the type of update: **Bug**, **Dependencies**, **Enhancement**, or **New-Feature**
- [x] Tag the PR with **Azure API for FHIR** if this will release to the managed service
- Review [squash-merge requirements](https://github.com/microsoft/fhir-server/blob/master/SquashMergeRequirements.md)

### Semver Change ([docs](https://github.com/microsoft/fhir-server/blob/master/docs/Versioning.md))
Patch|Skip|Feature|Breaking (reason)

### CODE REVIEW

**Comment by brendankowitz** [comment]
@feordin, @namadabu can you review since you have done work with background jobs

File: src/Microsoft.Health.Fhir.Core/Exceptions/RetriableTaskException.cs [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+using Microsoft.Health.Fhir.Core.Models;
+
+namespace Microsoft.Health.Fhir.Core.Exceptions
+{
+    public class RetriableTaskException : FhirException
+    {
+        public RetriableTaskException(string message)
+            : base(message)
+        {
+            Issues.Add(new OperationOutcomeIssue(
```
Comment (namadabu) [inline src/Microsoft.Health.Fhir.Core/Exceptions/RetriableTaskException.cs:15]: Are we planning on returning this information back to the user? I don't think these exceptions need to be of type `FhirException` unless we plan to return some of this information back to user.

File: src/Microsoft.Health.Fhir.Core/Features/TaskManagement/TaskInfo.cs [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+using System;
+
+namespace Microsoft.Health.Fhir.Core.Features.TaskManagement
+{
+    public class TaskInfo
+    {
+        public string TaskId { get; set; }
+
+        public string QueueId { get; set; }
```
Comment (namadabu) [inline src/Microsoft.Health.Fhir.Core/Features/TaskManagement/TaskInfo.cs:14]: Just for my understanding, do we expect there will be multiple queues for each fhir-server instance?

File: test/Microsoft.Health.Fhir.Shared.Tests.Integration/Persistence/SqlServerTaskConsumerTests.cs [hunk 3]
```diff
@@ -0,0 +1,373 @@
...
+        {
+            string queueId = Guid.NewGuid().ToString();
+            TaskHostingConfiguration config = new TaskHostingConfiguration()
+            {
+                Enabled = true,
+                QueueId = queueId,
+                TaskHeartbeatTimeoutThresholdInSeconds = 60,
+            };
+
+            IOptions<TaskHostingConfiguration> optionsExportConfig = Substitute.For<IOptions<TaskHostingConfiguration>>();
```
Comment (namadabu) [inline test/Microsoft.Health.Fhir.Shared.Tests.Integration/Persistence/SqlServerTaskConsumerTests.cs:44]: >optionsExportConfig [](start = 47, length = 19)

nit: maybe rename to `taskHostingConfig` ?

File: src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/9.diff.sql [hunk 4]
```diff
@@ -0,0 +1,441 @@
...
+	-- We will timestamp the jobs when we update them to track stale jobs.
+    DECLARE @heartbeatDateTime datetime2(7) = SYSUTCDATETIME()
+
+    IF (@retryCount IS NULL) BEGIN
+        THROW 50404, 'Task not exist or runid not match', 1;
+    END
+
+    IF (@retryCount >= @maxRetryCount) BEGIN
+		UPDATE dbo.TaskInfo
+		SET Status = 3, HeartbeatDateTime = @heartbeatDateTime, Result = @result
```
Comment (namadabu) [inline src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/9.diff.sql:364]: >Result = @result [](start = 58, length = 16)

Should we update the `Result` here even though we are beyond `maxRetryCount`?

File: src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/9.diff.sql [hunk 5]
```diff
@@ -0,0 +1,441 @@
...
+	DECLARE @status smallint = 1
+    DECLARE @retryCount smallint = 0
+	DECLARE @isCanceled bit = 0
+
+    -- Check if the task already be created
+    DECLARE @existedTaskCount bigint
+
+    SELECT @existedTaskCount = Count(*)
+    FROM [dbo].[TaskInfo]
+    WHERE TaskId = @taskId
```
Comment (namadabu) [inline src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/9.diff.sql:70]: Is this necessary since we have another check below?

File: src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/9.diff.sql [hunk 6]
```diff
@@ -0,0 +1,441 @@
...
+        THROW 50409, 'Task already existed', 1;
+    END
+
+    -- Create new task
+    INSERT INTO [dbo].[TaskInfo]
+        (TaskId, QueueId, Status, TaskTypeId, IsCanceled, RetryCount, HeartbeatDateTime, InputData)
+    VALUES
+        (@taskId, @queueId, @status, @taskTypeId, @isCanceled, @retryCount, @heartbeatDateTime, @inputData)
+
+    SELECT TaskId, QueueId, Status, TaskTypeId, RunId, IsCanceled, RetryCount, HeartbeatDateTime, InputData, TaskContext, Result
```
Comment (namadabu) [inline src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/9.diff.sql:87]: >TaskContext, Result [](start = 109, length = 19)

nit: Do we need to return `TaskContext` and `Result` since they are always `NULL` here?

File: src/Microsoft.Health.Fhir.Core/Features/TaskManagement/TaskHosting.cs [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+namespace Microsoft.Health.Fhir.Core.Features.TaskManagement
+{
+    public class TaskHosting
+    {
+        private object _syncRoot = new object();
+
+        private ITaskConsumer _consumer;
+        private ITaskFactory _taskFactory;
+        private ILogger<TaskHosting> _logger;
+        private Dictionary<string, ITask> _activeTaskRecordsForKeepAlive = new Dictionary<string, ITask>();
```
Comment (feordin) [inline src/Microsoft.Health.Fhir.Core/Features/TaskManagement/TaskHosting.cs:23]: It might be possible to use a ConcurrentDictionary here to avoid the need of a lock object.

**Comment by feordin** [comment]
Will it be possible for tasks to be shared across multiple FhirService instances? Or will a task be handled only by one instance?

File: src/Microsoft.Health.Fhir.Core/Features/TaskManagement/ITask.cs [hunk 8]
```diff
@@ -0,0 +1,21 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.Threading.Tasks;
+
+namespace Microsoft.Health.Fhir.Core.Features.TaskManagement
```
Comment (namadabu) [inline src/Microsoft.Health.Fhir.Core/Features/TaskManagement/ITask.cs:9]: I don't think there is anything FHIR specific to all these Task related classes. It seems to me that instead of them being in Fhir.Core, it should be a separate project. And that will also keep things modular so that if other users want to use a different task management mechanism they can easily insert their own implementation.

File: src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/9.diff.sql (hunk 4 above)
Comment (tongwu-msft) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/9.diff.sql:364]: Update result here is for the store the failed task result. Client would get failed result details for the last retry.

File: src/Microsoft.Health.Fhir.Core/Features/TaskManagement/TaskInfo.cs (hunk 2 above)
Comment (tongwu-msft) [reply src/Microsoft.Health.Fhir.Core/Features/TaskManagement/TaskInfo.cs:14]: Yes, we want to support multiple queue. One typical scénario is like we have a orchestrator task and multiple sub tasks. Orchestrator task need high availability and less resource it is to create and monitor sub task. And sub task is highly resource consumer. We'd better have 2 different environment with different capacity. We can use different queueid to dispatch task to different environment. Currently we have only one queueid in FHIR server, and we plan to enable it in separate runtime environment.

File: test/Microsoft.Health.Fhir.Shared.Tests.Integration/Persistence/SqlServerTaskConsumerTests.cs (hunk 3 above)
Comment (tongwu-msft) [reply test/Microsoft.Health.Fhir.Shared.Tests.Integration/Persistence/SqlServerTaskConsumerTests.cs:44]: Thanks, updated.

File: src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/9.diff.sql (hunk 5 above)
Comment (tongwu-msft) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/9.diff.sql:70]: Yes, it is duplicated.

File: src/Microsoft.Health.Fhir.Core/Features/TaskManagement/TaskHosting.cs (hunk 7 above)
Comment (tongwu-msft) [reply src/Microsoft.Health.Fhir.Core/Features/TaskManagement/TaskHosting.cs:23]: Looks like it would not have conflict on same key write, updated to ConsurrentDictionary.

**Comment by tongwu-msft** [comment]
> Will it be possible for tasks to be shared across multiple FhirService instances? Or will a task be handled only by one instance?

Since we might have multiple instances and there're lot of possible unexcepted error like server crash/hang... It is hard to guarantee there're no 2 running same tasks all the time. Our solution is to make sure for most cases task could not be repicked up when it is running. For the rare cases like the old running task hang/loss keepalive request but finally recover, we can detect and cancel the old running task. Details below:

- GetNextTasks would not return the task with recent keepalive. (it is running). 
- New runid generated at GetNextTasks, old task keepalive with old runid would get error and cancel itself.

File: src/Microsoft.Health.Fhir.Core/Features/TaskManagement/ITask.cs (hunk 8 above)
Comment (tongwu-msft) [reply src/Microsoft.Health.Fhir.Core/Features/TaskManagement/ITask.cs]: Agree, moved to a separated project.

File: src/Microsoft.Health.Fhir.Core/Exceptions/RetriableTaskException.cs (hunk 1 above)
Comment (tongwu-msft) [reply src/Microsoft.Health.Fhir.Core/Exceptions/RetriableTaskException.cs]: We do not expect customer see this exception, move to task management project.

File: src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/9.diff.sql (hunk 6 above)
Comment (tongwu-msft) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/9.diff.sql]: removed for unnecessary column

**Comment by tongwu-msft** [comment]
> @tongwu-msft Can you please add more context to the PR description regarding how we expect to use this infrastructure? It would be helpful to understand what kind of information is expected to be present in each task. I am assuming this is for import, but do we expect to use this for other scenarios in the future as well?

Update some sample for the usage, and here's the bulk import tasks implementation for more details: https://github.com/microsoft/fhir-server/tree/features/bulkimport%2Fdev

**Review by feordin — APPROVED** [review]

File: src/Microsoft.Health.Fhir.SqlServer/Microsoft.Health.Fhir.SqlServer.csproj [hunk 9]
```diff
@@ -70,12 +78,12 @@
       <Generated Include="Features\Schema\Model\ImmutableTypes.Generated.cs">
         <Generator>ImmutableSqlModelGenerator</Generator>
         <Namespace>Microsoft.Health.Fhir.SqlServer.Features.Schema.Model</Namespace>
-        <Args>@(ImmutableSqlGeneratorInputs->'&quot;%(FullPath)&quot;', ' ')</Args>
+        <Args>@(ImmutableSqlGeneratorInputs->'"%(FullPath)"', ' ')</Args>
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.SqlServer/Microsoft.Health.Fhir.SqlServer.csproj:81]: Is this change intentional?

File: src/Microsoft.Health.Fhir.TaskManagement.UnitTests/TestTaskConsumer.cs [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+                    t.Status = TaskStatus.Queued;
+                }
+
+                t.HeartbeatDateTime = DateTime.Now;
+            }
+        }
+
+        public Task<TaskInfo> CompleteAsync(string taskId, TaskResultData result, string runId, CancellationToken cancellationToken)
+        {
+            _faultInjectionAction?.Invoke("CompleteAsync");
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.TaskManagement.UnitTests/TestTaskConsumer.cs:40]: Can this use `nameof()`?

File: src/Microsoft.Health.Fhir.TaskManagement/IContextUpdater.cs [hunk 11]
```diff
@@ -0,0 +1,14 @@
...
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System.Threading.Tasks;
+
+namespace Microsoft.Health.Fhir.TaskManagement
+{
+    public interface IContextUpdater
+    {
+        public Task<TaskInfo> UpdateContextAsync(string taskId, string context);
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.TaskManagement/IContextUpdater.cs:12]: Can you add docs for this?

File: src/Microsoft.Health.Fhir.TaskManagement/ITask.cs [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.Threading.Tasks;
+
+namespace Microsoft.Health.Fhir.TaskManagement
+{
+    public interface ITask : IDisposable
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.TaskManagement/ITask.cs:11]: Lets add docs for the interfaces to help in future

File: src/Microsoft.Health.Fhir.TaskManagement/ITaskConsumer.cs [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Health.Fhir.TaskManagement
+{
+    public interface ITaskConsumer
+    {
+        Task<TaskInfo> CompleteAsync(string taskId, TaskResultData result, string runId, CancellationToken cancellationToken);
+
+        Task<IReadOnlyCollection<TaskInfo>> GetNextMessagesAsync(short count, int taskHeartbeatTimeoutThresholdInSeconds, CancellationToken cancellationToken);
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.TaskManagement/ITaskConsumer.cs:16]: Can `taskHeartbeatTimeoutThresholdInSeconds` just be a timespan?

File: src/Microsoft.Health.Fhir.TaskManagement/RetriableTaskException.cs [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+
+using System;
+
+namespace Microsoft.Health.Fhir.TaskManagement
+{
+    public class RetriableTaskException : Exception
+    {
+        public RetriableTaskException(string message)
+            : base(message)
+        {
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.TaskManagement/RetriableTaskException.cs:14]: Assert message not null

File: src/Microsoft.Health.Fhir.TaskManagement/TaskHosting.cs [hunk 15]
```diff
@@ -0,0 +1,201 @@
...
+    public class TaskHosting
+    {
+        private ITaskConsumer _consumer;
+        private ITaskFactory _taskFactory;
+        private ILogger<TaskHosting> _logger;
+        private ConcurrentDictionary<string, ITask> _activeTaskRecordsForKeepAlive = new ConcurrentDictionary<string, ITask>();
+
+        public TaskHosting(ITaskConsumer consumer, ITaskFactory taskFactory, ILogger<TaskHosting> logger)
+        {
+            _consumer = consumer;
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.TaskManagement/TaskHosting.cs:24]: EnsureArg.Isnotnull

File: src/Microsoft.Health.Fhir.TaskManagement/TaskHosting.cs [hunk 16]
```diff
@@ -0,0 +1,201 @@
...
+                        runningTasks.Add(ExecuteTaskAsync(taskInfo, cancellationToken));
+                    }
+                }
+
+                await intervalDelayTask;
+            }
+
+            try
+            {
+                Task.WaitAll(runningTasks.ToArray());
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.TaskManagement/TaskHosting.cs:87]: does this need an `await`?

File: src/Microsoft.Health.Fhir.TaskManagement/TaskHosting.cs [hunk 17]
```diff
@@ -0,0 +1,201 @@
...
+            }
+            catch (Exception ex)
+            {
+                _logger.LogError(ex, "Task failed to execute");
+            }
+        }
+
+        private async Task ExecuteTaskAsync(TaskInfo taskInfo, CancellationToken cancellationToken)
+        {
+            using ITask task = _taskFactory.Create(taskInfo);
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.TaskManagement/TaskHosting.cs:97]: EnsureArg for taskInfo

File: src/Microsoft.Health.Fhir.TaskManagement/TaskHosting.cs [hunk 18]
```diff
@@ -0,0 +1,201 @@
...
+            task.RunId = task.RunId ?? taskInfo.RunId;
+
+            if (task == null)
+            {
+                _logger.LogWarning("Not supported task type: {0}", taskInfo.TaskTypeId);
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.TaskManagement/TaskHosting.cs:102]: nit: Use structured logging variable
```suggestion
                _logger.LogWarning("Not supported task type: {taskTypeId}", taskInfo.TaskTypeId);
```

File: src/Microsoft.Health.Fhir.TaskManagement/TaskHosting.cs [hunk 19]
```diff
@@ -0,0 +1,201 @@
...
+                            _logger.LogError(notExistEx, "Task {0} not exist or runid not match.", taskId);
+                            shouldCancel = true;
+                        }
+
+                        if (shouldCancel)
+                        {
+                            task.Cancel();
+                        }
+                    }
+                    catch (Exception ex)
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.TaskManagement/TaskHosting.cs:189]: Can we catch a specific exception instead of being generic? i.e. `TaskCancelledException`

File: src/Microsoft.Health.Fhir.TaskManagement/TaskNotExistException.cs [hunk 20]
```diff
@@ -0,0 +1,22 @@
...
+
+using System;
+
+namespace Microsoft.Health.Fhir.TaskManagement
+{
+    public class TaskNotExistException : Exception
+    {
+        public TaskNotExistException(string message)
+            : base(message)
+        {
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.TaskManagement/TaskNotExistException.cs:14]: nit: Assert message not null

File: src/Microsoft.Health.Fhir.TaskManagement/TaskResultData.cs [hunk 21]
```diff
@@ -0,0 +1,20 @@
...
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+namespace Microsoft.Health.Fhir.TaskManagement
+{
+    public class TaskResultData
+    {
+        public TaskResultData(TaskResult result, string resultData)
+        {
+            Result = result;
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.TaskManagement/TaskResultData.cs:12]: EnsureArg.NotNull

File: src/Microsoft.Health.Fhir.TaskManagement/TaskStatus.cs [hunk 22]
```diff
@@ -0,0 +1,15 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+namespace Microsoft.Health.Fhir.TaskManagement
+{
+    public enum TaskStatus
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.TaskManagement/TaskStatus.cs:8]: Document enums

File: test/Microsoft.Health.Fhir.Shared.Tests.Integration/Persistence/SqlServerTaskConsumerTests.cs [hunk 23]
```diff
@@ -0,0 +1,373 @@
...
+    {
+        private SqlServerFhirStorageTestsFixture _fixture;
+
+        public SqlServerTaskConsumerTests(SqlServerFhirStorageTestsFixture fixture)
+        {
+            _fixture = fixture;
+        }
+
+        [Fact]
+        public async Task GivenListOfTasksInQueue_WhenGetNextTask_AvailableTasksShouldBeReturned()
```
Comment (brendankowitz) [inline test/Microsoft.Health.Fhir.Shared.Tests.Integration/Persistence/SqlServerTaskConsumerTests.cs:34]: nit: Convention is Given...When...Then... (think arrange, act, assert)

File: Microsoft.Health.Fhir.sln [hunk 24]
```diff
@@ -137,6 +137,10 @@ Project("{D954291E-2A0B-460D-934E-DC6B0785DB48}") = "Microsoft.Health.Fhir.Share
 EndProject
 Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "LogsAndInterpolations", "tools\FhirAnalyzer\FhirAnalyzer.csproj", "{389C844A-4DC2-4780-9821-464580396850}"
 EndProject
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Microsoft.Health.Fhir.TaskManagement", "src\Microsoft.Health.Fhir.TaskManagement\Microsoft.Health.Fhir.TaskManagement.csproj", "{03AD4FC4-A56F-4E94-B295-B02A9E3E3CCD}"
```
Comment (brendankowitz) [inline Microsoft.Health.Fhir.sln:140]: If this logic doesn't have anything to do with FHIR then it could just be `Microsoft.Health.TaskManagement` and sounds like it could be something that is later moved to Shared Components

File: src/Microsoft.Health.Fhir.SqlServer/Microsoft.Health.Fhir.SqlServer.csproj (hunk 9 above)
Comment (tongwu-msft) [reply src/Microsoft.Health.Fhir.SqlServer/Microsoft.Health.Fhir.SqlServer.csproj:81]: No, it changed automatically, fixed. Just curious why this need escape?

File: src/Microsoft.Health.Fhir.TaskManagement.UnitTests/TestTaskConsumer.cs (hunk 10 above)
Comment (tongwu-msft) [reply src/Microsoft.Health.Fhir.TaskManagement.UnitTests/TestTaskConsumer.cs]: that works, thanks

File: src/Microsoft.Health.Fhir.TaskManagement/IContextUpdater.cs (hunk 11 above)
Comment (tongwu-msft) [reply src/Microsoft.Health.Fhir.TaskManagement/IContextUpdater.cs:21]: Sure.

File: src/Microsoft.Health.Fhir.TaskManagement/TaskStatus.cs (hunk 22 above)
Comment (tongwu-msft) [reply src/Microsoft.Health.Fhir.TaskManagement/TaskStatus.cs:11]: Sure, updated.

File: src/Microsoft.Health.Fhir.TaskManagement/TaskResultData.cs (hunk 21 above)
Comment (tongwu-msft) [reply src/Microsoft.Health.Fhir.TaskManagement/TaskResultData.cs:16]: Sure, added.

File: src/Microsoft.Health.Fhir.TaskManagement/TaskNotExistException.cs (hunk 20 above)
Comment (tongwu-msft) [reply src/Microsoft.Health.Fhir.TaskManagement/TaskNotExistException.cs:15]: Sure, added.

File: src/Microsoft.Health.Fhir.TaskManagement/TaskHosting.cs (hunk 15 above)
Comment (tongwu-msft) [reply src/Microsoft.Health.Fhir.TaskManagement/TaskHosting.cs:29]: Sure added.

File: src/Microsoft.Health.Fhir.TaskManagement/RetriableTaskException.cs (hunk 14 above)
Comment (tongwu-msft) [reply src/Microsoft.Health.Fhir.TaskManagement/RetriableTaskException.cs:15]: Sure addded.

File: src/Microsoft.Health.Fhir.TaskManagement/TaskHosting.cs (hunk 17 above)
Comment (tongwu-msft) [reply src/Microsoft.Health.Fhir.TaskManagement/TaskHosting.cs:102]: sure, updated.

File: src/Microsoft.Health.Fhir.TaskManagement/TaskHosting.cs (hunk 18 above)
Comment (tongwu-msft) [reply src/Microsoft.Health.Fhir.TaskManagement/TaskHosting.cs:107]: thanks for the suggestion, updated.

File: src/Microsoft.Health.Fhir.TaskManagement/TaskHosting.cs (hunk 16 above)
Comment (tongwu-msft) [reply src/Microsoft.Health.Fhir.TaskManagement/TaskHosting.cs:92]: WhenAll looks better here. Updated.

File: src/Microsoft.Health.Fhir.TaskManagement/ITask.cs (hunk 12 above)
Comment (tongwu-msft) [reply src/Microsoft.Health.Fhir.TaskManagement/ITask.cs:14]: Sure, added.

File: src/Microsoft.Health.Fhir.TaskManagement/TaskHosting.cs (hunk 19 above)
Comment (tongwu-msft) [reply src/Microsoft.Health.Fhir.TaskManagement/TaskHosting.cs:196]: We catch general exception here is to prevent any possible exception throw here to break the keep alive thread. Specific exception would be handled inside.

File: src/Microsoft.Health.Fhir.TaskManagement/ITaskConsumer.cs (hunk 13 above)
Comment (tongwu-msft) [reply src/Microsoft.Health.Fhir.TaskManagement/ITaskConsumer.cs:34]: For this threshold we target to have reasonable level precision and looks like seconds is enough (keepalive is at seconds level), not sure is there any special case we might need general timespan?

File: Microsoft.Health.Fhir.sln (hunk 24 above)
Comment (tongwu-msft) [reply Microsoft.Health.Fhir.sln]: Updated to taskmanagement.

File: test/Microsoft.Health.Fhir.Shared.Tests.Integration/Persistence/SqlServerTaskConsumerTests.cs (hunk 23 above)
Comment (tongwu-msft) [reply test/Microsoft.Health.Fhir.Shared.Tests.Integration/Persistence/SqlServerTaskConsumerTests.cs]: thanks, fixed.
